### PR TITLE
Converting DateTime.MinValue to sqlDateTime's minimum value

### LINKED ISSUE #19865: v16: Time-only DateTime picker
### Which Umbraco version are you using?

16.1.1

### Bug summary

I want to have a time-only property editor. Everything in the backoffice facilitates this and you even get a nice time-only picker, but it will throw an error when saving.

### Specifics

_No response_

### Steps to reproduce

Create a time data type using the Date Picker Property Editor UI and supply a date format of HH:mm:

Image

When you use that data type on your content type, it actually displays a very nice time selector that even has a clock instead of a calendar icon:

Image

It even supports the pickers to select the time:

Image

Awesome, that's exactly when I want! However when saving....

Image

And it's the obvious error I expected:
System.Data.SqlTypes.SqlTypeException: SqlDateTime overflow. Must be between 1/1/1753 12:00:00 AM and 12/31/9999 11:59:59 PM.

### Expected result / actual result

I guess it's a scenario that wasn't considered, but since it works so well to have a time-only property editor, it's a real shame it cannot be used. I think this should be fixed one of two ways:

* Support the time-only version (can't we use datetime2 in the database?)
* Or make sure that the date format on the property editor cannot be invalid. (In which case I need to make a custom data editor for time-only myself...).

---
_This item has been added to our backlog AB#55868_

### PR DESCRIPTION
**Pull request by NillasKA**
If there's an existing issue for this PR then this fixes [](https://github.com/umbraco/Umbraco-CMS/issues/19865)

### Description
A bug was discovered, whereas if you a date property editor with a time only format, any documents created with this would be unable to save.

It turns out, that when only a time has been set, and no date is ever picked, we recieve the date 01/01/0001, which is `DateTime.MinValue` - Our issue here is that SQLServer's DateTime datatype has a minimum value of 01/01/1753, so anything below that will cause issues.

I have just made a small check in the value converter to check if the date we are receiving is 01/01/0001, if it is, we just

Closes #19865

### CODE REVIEW

**Comment by NillasKA** [comment]
Fun read as to why the sqlDateTime value minimum is the year 1753, for anyone curios: https://stackoverflow.com/questions/3310569/what-is-the-significance-of-1-1-1753-in-sql-server

So this of course would not be a problem if we were using datetime2 as our datatype in our sql database, but because we are using the sqlDateTime datatype, we run into this odd decision made years ago by the nice folks behind the old SQL Server

File: src/Umbraco.Core/Extensions/ObjectExtensions.cs [hunk 1]
```diff
@@ -491,6 +491,14 @@ internal static void CheckThrowObjectDisposed(this IDisposable disposable, bool
         {
             if (DateTime.TryParse(input, out DateTime value))
             {
+                 // Minimum value of SQL Servers sqlDateTime property is 01/01/1753.
+                 // So if a DateTime.MinValue (01/01/0001) comes in, we need to convert it to sqlDateTime min value.
+                 // This can be removed if we ever migrate our property to DateTime2
+                if (value.Year < 1753)
+                {
+                    value = new DateTime(1753, value.Month, value.Day, value.Hour, value.Minute, value.Second, value.Millisecond);
+                }
```
Comment (bjarnef) [inline src/Umbraco.Core/Extensions/ObjectExtensions.cs]: Can it use this property instead of the hardcoded value?
https://learn.microsoft.com/en-us/dotnet/api/system.data.sqltypes.sqldatetime.minvalue?view=net-9.0

Comment (NillasKA) [reply src/Umbraco.Core/Extensions/ObjectExtensions.cs]: Good shout - I have adjusted accordingly

**Review by AndyButland — CHANGES_REQUESTED** [review]
Unfortunately I don't think we can make the update here, as it's in a very general function for converting objects to particular types, and we can't know how implementors are using it.  It's possible someone is working with dates below the SQL Server minimum - just presumably not storing them in SQL Server - and if we change this, we'd break the behaviour for them.

We only want this to apply where we are actually looking to store values that have come from property editors, so it may be better placed just before the `return value.TryConvertTo(valueType);` line in `DataEditor`.

Or better still, we could introduce a custom value editor for the date picker property editor, that would apply this change only there.  If you look at `IntegerPropertyEditor` you'll see an example there where a custom value editor is registered and the object is amended going to and from storage in the `ToEditor` and `FromEditor` methods.  So we could do similar in `DateTimePropertyEditor`.  It should then be possible to check that we have a configured format that matches a time-only one, and only apply this conversion then.  That way we only do it when it's necessary and useful.

Bear in mind as I said before we - well Laura - is working on a new property editor that would better support time-only, so we could wait for that.  But I still think if we can provide this tightly in the existing one it could be valuable for people, so it's worth looking at a bit more if you are up for it.

File: src/Umbraco.Core/Models/Editors/ContentPropertyData.cs [hunk 2]
```diff
@@ -18,7 +18,7 @@ public ContentPropertyData(object? value, object? dataTypeConfiguration)
     /// <summary>
     ///     The value submitted for the property
     /// </summary>
-    public object? Value { get; }
+    public object? Value { get; set; }
```
Comment (AndyButland) [inline src/Umbraco.Core/Models/Editors/ContentPropertyData.cs]: I don't think you need this change (see comment below), so can revert.

File: src/Umbraco.Infrastructure/PropertyEditors/DateTimePropertyEditor.cs [hunk 3]
```diff
@@ -25,8 +29,45 @@ public DateTimePropertyEditor(IDataValueEditorFactory dataValueEditorFactory)
...
     protected override IDataValueEditor CreateValueEditor()
     {
-        IDataValueEditor editor = base.CreateValueEditor();
-        editor.Validators.Add(new DateTimeValidator());
-        return editor;
+        return DataValueEditorFactory.Create<DateTimePropertyValueEditor>(Attribute!);
+
+        // IDataValueEditor editor = base.CreateValueEditor();
+        // editor.Validators.Add(new DateTimeValidator());
+        // return editor;
```
Comment (AndyButland) [inline src/Umbraco.Infrastructure/PropertyEditors/DateTimePropertyEditor.cs]:
```suggestion
        DateTimePropertyValueEditor editor = DataValueEditorFactory.Create<DateTimePropertyValueEditor>(Attribute!);
        editor.Validators.Add(new DateTimeValidator());
        return editor;
```

We should remove the old commented out code here, but we do need to retain the  validator.

File: src/Umbraco.Infrastructure/PropertyEditors/DateTimePropertyEditor.cs [hunk 4]
```diff
@@ -25,8 +29,45 @@ public DateTimePropertyEditor(IDataValueEditorFactory dataValueEditorFactory)
...
+                // Ensures we don't care about the customers formatting, weather the wish to write
+                // hh:mm or HH:MM etc...
+                if (!dictionary.Values.OfType<string>().Any(value => array.Contains(value, StringComparer.InvariantCultureIgnoreCase)))
+                {
+                    return editorValue.Value;
+                }
+
+                if (DateTime.TryParse(editorValue.Value as string, out DateTime myDate) || editorValue.Value is DateTime)
+                {
+                    editorValue.Value = new DateTime(SqlDateTime.MinValue.Value.Year, myDate.Month, myDate.Day, myDate.Hour, myDate.Minute, myDate.Second).ToString("yyyy-MM-dd HH:mm:ss");
```
Comment (AndyButland) [inline src/Umbraco.Infrastructure/PropertyEditors/DateTimePropertyEditor.cs]:
```suggestion
                    return new DateTime(SqlDateTime.MinValue.Value.Year, myDate.Month, myDate.Day, myDate.Hour, myDate.Minute, myDate.Second).ToString("yyyy-MM-dd HH:mm:ss");
```

This avoids having to make the change to `ContentPropertyData.Value`.

File: tests/Umbraco.Tests.UnitTests/Umbraco.Infrastructure/PropertyEditors/DateTimePropertyEditorTests.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace Umbraco.Cms.Tests.UnitTests.Umbraco.Infrastructure.PropertyEditors;
+
+public class DateTimePropertyEditorTests
+{
+[Test]
+[TestCase("01/01/0001 10:00", true)]
+[TestCase("10/10/1000 10:00", true)]
+[TestCase("01/01/2000 10:00", false)]
+public void Ensure_Correct_DateTime_Conversion(DateTime dateTime, bool shouldConvert)
```
Comment (AndyButland) [inline tests/Umbraco.Tests.UnitTests/Umbraco.Infrastructure/PropertyEditors/DateTimePropertyEditorTests.cs]:
```suggestion
    [TestCase("01/01/0001 10:00", true)]
    [TestCase("10/10/1000 10:00", true)]
    [TestCase("01/01/2000 10:00", false)]
    public void Ensure_Correct_DateTime_Conversion(DateTime dateTime, bool shouldConvert)
```

Just fixed indentation and removed the unnecessary `[Test]`, which isn't needed if you have the `[TestCase` ones,

Comment (AndyButland) [inline tests/Umbraco.Tests.UnitTests/Umbraco.Infrastructure/PropertyEditors/DateTimePropertyEditorTests.cs]: I'd suggest reworking this test slightly, so rather than passing in `shouldConvert`, which isn't super-clear from the naming, provide instead an expected date time (which will have to be provided as a string) and assert against that.

You could also expand it further with more test cases, by providing the format to the test method too as a parameter, so you can test that the adjustment of the minimum date only happens when you are using a time format.

File: src/Umbraco.Infrastructure/PropertyEditors/DateTimePropertyEditor.cs [hunk 6]
```diff
@@ -25,8 +29,45 @@ public DateTimePropertyEditor(IDataValueEditorFactory dataValueEditorFactory)
...
+        /// <inheritdoc/>
+        public override object? FromEditor(ContentPropertyData editorValue, object? currentValue)
+        {
+            if (editorValue.DataTypeConfiguration is Dictionary<string, object> dictionary)
+            {
+                string[] array = ["hh:mm", "hh:mm:ss"];
+
+                // Ensures we don't care about the customers formatting, weather the wish to write
+                // hh:mm or HH:MM etc...
+                if (!dictionary.Values.OfType<string>().Any(value => array.Contains(value, StringComparer.InvariantCultureIgnoreCase)))
```
Comment (AndyButland) [inline src/Umbraco.Infrastructure/PropertyEditors/DateTimePropertyEditor.cs]: This needs tightening up a little I think.

Firstly, can we check the specific configuration dictionary entry that contains the format string, rather than all of them?  It's highly unlikely that any other field will contain a format string, but you never know what could be added in future.

Secondly, we need to check both more and less of the format strings!  Case sensitivity is important here as "M" represents month and "m" minute.  So when you say the customer could format "HH:MM" this would be incorrect.  But they could have "hh-mm", or "h-mm" or "hh : mm", or anything they want really - so there are more variations.

I think the way to do it is to find all the unique characters in the format string, and consider it time only if the only characters found are "s", "m" or "h".  That'll be a fun algorithm to write (or you can lean on our digital overlords).

P.S. You'll probably change the comment anyway in amending this, but "weather the write" --> "whether they right".

File: src/Umbraco.Infrastructure/PropertyEditors/DateTimePropertyEditor.cs [hunk 7]
```diff
@@ -25,8 +29,45 @@ public DateTimePropertyEditor(IDataValueEditorFactory dataValueEditorFactory)
...
+                    return editorValue.Value;
+                }
+
+                if (DateTime.TryParse(editorValue.Value as string, out DateTime myDate) || editorValue.Value is DateTime)
+                {
+                    editorValue.Value = new DateTime(SqlDateTime.MinValue.Value.Year, myDate.Month, myDate.Day, myDate.Hour, myDate.Minute, myDate.Second).ToString("yyyy-MM-dd HH:mm:ss");
+                }
+            }
+
+            return editorValue.Value;
```
Comment (AndyButland) [inline src/Umbraco.Infrastructure/PropertyEditors/DateTimePropertyEditor.cs]: If we don't do any amends for time formats, we are just returning `editorValue.Value` (here, and also above).  However this might be different to what happens had we not introduced our custom value editor, and instead, as previously, the default `DataValueEditor` was used.

To be sure we use that same functionality, it's probably better to return `base.FromEditor(....` in this situation.

**Review by AndyButland — CHANGES_REQUESTED** [review]
As mentioned last week @NillasKA, I think this is the right approach to take here.  But I've found a few things to consider, which I'll put in inline comments.

**Review by AndyButland — APPROVED** [review]
Looks good to me now @NillasKA.  I pushed a few changes - some tidy up, a fix from testing, but mostly around the detection of the time format.  I figured it would be more accurate if we relied on date parsing rather than trying to detect all the different characters and separators.

I've done some testing as I say, but please have a further check over by creating date pickers of different formats, and make sure you can add and update them in the backoffice and persist them in templates on the front-end.  As I say, it looks good to me, and we've not got this update quite tight so it only affects people who are trying to use the date picker for time only, but we need to be careful here.

Please also of course check over the updates I made and make sure you are happy with them.

After that though, please go ahead and merge.

When you do, you should close the associated issue with a comment saying "Resolved for 16.3 in {link to this PR}.".  Label the issue with `release/16.3` and remove the two `state/*` labels.

**Comment by NillasKA** [comment]
lgtm
